school/database.py

```diff
@@ -0,0 +1,514 @@
+        The function must return Boolean.
+        """
+        if func is None:
+            return
+
+        if func not in Teacher._is_used_checks:
+            Teacher._is_used_checks.append(func)
+
+    @staticmethod
+    def get_by_sid(ctx, school_id: int) -> Optional[Teacher]:
+        query = (
+            session.query(Teacher)
+            .filter_by(guild_id=ctx.guild.id)
+            .filter_by(school_id=school_id)
+        )
+
+        return query.one_or_none()
+
+    @staticmethod
+    def search(ctx, name: str) -> List[Teacher]:
+        query = (
+            session.query(Teacher)
+            .filter_by(guild_id=ctx.guild.id)
+            .filter(Teacher.name.ilike(f"%{name}%"))
+        )
+        return query.all()
+
+    def edit(self, name):
+        if name:
+            self.name = name
+        session.commit()
+
+    @staticmethod
+    def get_or_create(ctx, school_id: int, name: str) -> Teacher:
+        query = session.query(Teacher).filter_by(school_id=school_id).one_or_none()
+
+        if query:
+            return query
+
+        teacher = Teacher(school_id=school_id, name=name, guild_id=ctx.guild.id)
+
+        session.add(teacher)
+        session.commit()
+
+        return teacher
+
+    def is_used(self) -> Boolean:
+        for func in Teacher._is_used_checks:
+            if func():
+                return True
+
+        return False
+
+    def delete(self):
+        session.delete(self)
+        session.commit()
+
+    @staticmethod
+    def purge():
+        """BEWARE! Purges all teachers that are not used!"""
+
+        query = session.query(Teacher)
+
+        for func in Teacher._used_filter_generators:
+            query = query.filter(not_(func()))
+
+        query.delete(synchronize_session="fetch")
+
+        session.commit()
+
+    def __repr__(self) -> str:
+        return (
+            f'<{self.__class__.__name__} idx="{self.idx}" school_id="{self.school_id}"'
+            f'guild_id="{self.guild_id}" name="{self.name}">'
+        )
+
+    def dump(self) -> dict:
+        return {
+            "idx": self.id,
+            "schoold_id": self.school_id,
+            "guild_id": self.guild_id,
+            "name": self.name,
+            "subjects": self.subjects,
+        }
+
+
+class Program(database.base):
+    __tablename__ = "school_dataset_program"
+    __table_args__ = (
+        UniqueConstraint(
+            "guild_id",
+            "abbreviation",
+            "degree",
+            name="guild_id_abbreviation_degree_unique",
+        ),
+    )
+
+    idx = Column(Integer, primary_key=True, autoincrement=True)
+    guild_id = Column(BigInteger)
+    abbreviation = Column(String)
+    name = Column(String)
+    degree = Column(String)
+    subjects = relationship(
+        "SubjectProgram",
+        back_populates="program",
+        cascade="all, delete-orphan, delete",
+    )
+
+    @staticmethod
+    def get_by_abbreviation(ctx, degree: str, abbreviation: str) -> Optional[Program]:
```

Comment:
I prefer using this because you don't have to check if the parameters are provided and it's more readable for other developers (they know what they get)

Reply:
I've though about it and I've changed it the way you said.